### PR TITLE
perf: tweak memory related check of AsmMachine

### PR DESCRIPTION
**Pull request by quake**
cargo bench "interpret secp256k1_bench via assembly" --features asm

5 times avg:
```
develop: 2.2431 ms
this pr: 2.2265 ms
```

### CODE REVIEW

File: src/machine/asm/mod.rs [hunk 1]
```diff
@@ -101,6 +102,97 @@ fn check_memory(machine: &mut AsmCoreMachine, page_indices: &(u64, u64)) {
...
+    addr: u64,
+    size: usize,
+) -> Result<(), Error> {
+    debug_assert!(size == 1 || size == 2 || size == 4 || size == 8);
+    let page = addr >> RISCV_PAGE_SHIFTS;
+    if page as usize >= RISCV_PAGESIZE {
+        return Err(Error::OutOfBound);
+    }
+
+    let page_offset = addr as usize % RISCV_PAGESIZE;
```
Comment (mohanson) [inline src/machine/asm/mod.rs:126]: May be faster if the compiler does not do this optimization : `let page_offset = addr as usize & 0xfff`

File: src/machine/asm/mod.rs [hunk 2]
```diff
@@ -101,6 +102,97 @@ fn check_memory(machine: &mut AsmCoreMachine, page_indices: &(u64, u64)) {
...
+        }
+
+        let frame = page >> MEMORY_FRAME_PAGE_SHIFTS;
+        if machine.frames[frame as usize] == 0 {
+            inited_memory(frame, machine);
+            machine.frames[frame as usize] = 1;
+        }
+
+        machine.set_flag(page, FLAG_DIRTY)?
+    } else {
```
Comment (mohanson) [inline src/machine/asm/mod.rs]: If a `write` method use two pages, then two pages should be checked, but you only check the latter page here.

File: src/machine/asm/mod.rs [hunk 3]
```diff
@@ -198,15 +290,17 @@ impl Memory for Box<AsmCoreMachine> {
...
-        let page_indices = get_page_indices(addr, 4)?;
-        check_permission(self, &page_indices, FLAG_EXECUTABLE)?;
-        self.load32(&(addr)).map(|v| v as u32)
+        check_memory_executable(self, addr, 4)?;
+        Ok(LittleEndian::read_u32(
+            &self.memory[addr as usize..addr as usize + 4],
+        ))
     }
 
     fn load8(&mut self, addr: &u64) -> Result<u64, Error> {
```
Comment (mohanson) [inline src/machine/asm/mod.rs:309]: It would be great if you can also modify load8/16/32/64

Comment (quake) [reply src/machine/asm/mod.rs:309]: resolved, please review again.

File: src/machine/asm/mod.rs (hunk 2 above)
Comment (quake) [reply src/machine/asm/mod.rs]: sorry for the mistake, please review again.

File: src/machine/asm/mod.rs (hunk 1 above)
Comment (quake) [reply src/machine/asm/mod.rs:126]: I can confirm that the compiler will optimize the code, will keep it unchanged

**Review by mohanson — APPROVED** [review]
LGTM

**Pull request merged**